File: bot/src/data/roast.ts
import type Database from "better-sqlite3";

export interface RoastQuote {
  id: number;
  userUid: string;
  userName: string;
  text: string;
  createdAt: number;
  /** LLM cringe score 0–10; null = not yet graded. */
  score: number | null;
  reason: string | null;
}

const META_LAST_ROAST_AT = "last_roast_at";

/**
 * Storage for the "roast" feature (ROADMAP Phase 8): capture each member's chat
 * lines → LLM cringe-grade (0–10 + one-line reason) → compile a greatest-hits
 * reel when enough people are present. Opting out purges a user's lines and
 * blocks further capture. Backed by the bot's existing SQLite db, so the MVP
 * needs no new infrastructure (the vector-DB/MemPalace upgrades come later).
 */
export class RoastStore {
  private addStmt: Database.Statement;
  private ungradedStmt: Database.Statement;
  private gradeStmt: Database.Statement;
  private topStmt: Database.Statement;
  private optInStmt: Database.Statement;
  private isOptedStmt: Database.Statement;
  private purgeStmt: Database.Statement;
  private metaGetStmt: Database.Statement;
  private metaSetStmt: Database.Statement;
  private countGradedStmt: Database.Statement;
  private countUngradedStmt: Database.Statement;
  private recentDupStmt: Database.Statement;
  private removeIdsStmt: Database.Statement;

  constructor(db: Database.Database) {
    db.exec(`
      CREATE TABLE IF NOT EXISTS roast_quotes (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        user_uid TEXT NOT NULL,
        user_name TEXT NOT NULL,
        text TEXT NOT NULL,
        created_at INTEGER NOT NULL,
        score INTEGER,
        reason TEXT
      );
      CREATE INDEX IF NOT EXISTS idx_roast_quotes_score ON roast_quotes(score);
      CREATE TABLE IF NOT EXISTS roast_optout (
        user_uid TEXT PRIMARY KEY
      );
      CREATE TABLE IF NOT EXISTS roast_meta (
        key TEXT PRIMARY KEY,
        value TEXT NOT NULL
      );
    `);
    this.addStmt = db.prepare(
      `INSERT INTO roast_quotes (user_uid, user_name, text, created_at) VALUES (?, ?, ?, ?)`,
    );
    this.ungradedStmt = db.prepare(
      `SELECT * FROM roast_quotes WHERE score IS NULL ORDER BY created_at ASC LIMIT ?`,
    );
    this.gradeStmt = db.prepare(`UPDATE roast_quotes SET score = ?, reason = ? WHERE id = ?`);
    this.topStmt = db.prepare(
      `SELECT * FROM roast_quotes WHERE score IS NOT NULL ORDER BY score DESC, created_at DESC LIMIT ?`,
    );
    this.optInStmt = db.prepare(`INSERT OR IGNORE INTO roast_optout (user_uid) VALUES (?)`);
    this.isOptedStmt = db.prepare(`SELECT 1 FROM roast_optout WHERE user_uid = ?`);
    this.purgeStmt = db.prepare(`DELETE FROM roast_quotes WHERE user_uid = ?`);
    this.metaGetStmt = db.prepare(`SELECT value FROM roast_meta WHERE key = ?`);
    this.metaSetStmt = db.prepare(
      `INSERT INTO roast_meta (key, value) VALUES (?, ?) ON CONFLICT(key) DO UPDATE SET value = excluded.value`,
    );
    this.countGradedStmt = db.prepare(
      `SELECT COUNT(*) AS n FROM roast_quotes WHERE score IS NOT NULL AND score >= ?`,
    );
    this.countUngradedStmt = db.prepare(
      `SELECT COUNT(*) AS n FROM roast_quotes WHERE score IS NULL`,
    );
    this.recentDupStmt = db.prepare(
      `SELECT 1 FROM roast_quotes WHERE user_uid = ? AND text = ? AND created_at > ? LIMIT 1`,
    );
    this.removeIdsStmt = db.prepare(`DELETE FROM roast_quotes WHERE id = ?`);
  }

  /** Record one captured line. Callers must check {@link isOptedOut} first. */
  add(userUid: string, userName: string, text: string): void {
    this.addStmt.run(userUid, userName, text, Date.now());
  }

  /** Skip near-duplicate lines from the same user (spam / echo). */
  hasRecentDuplicate(userUid: string, text: string, withinMs: number): boolean {
    const since = Date.now() - withinMs;
    return !!this.recentDupStmt.get(userUid, text, since);
  }

  /** Oldest ungraded lines, for a batched grading pass. */
  ungraded(limit = 25): RoastQuote[] {
    return (this.ungradedStmt.all(limit) as unknown[]).map(rowToQuote);
  }

  ungradedCount(): number {
    return (this.countUngradedStmt.get() as { n: number }).n;
  }

  setGrade(id: number, score: number, reason: string): void {
    this.gradeStmt.run(clampScore(score), reason.slice(0, 280), id);
  }

  /** Highest-graded lines for the compilation. */
  top(limit = 8): RoastQuote[] {
    return (this.topStmt.all(limit) as unknown[]).map(rowToQuote);
  }

  gradedCount(minScore = 0): number {
    return (this.countGradedStmt.get(minScore) as { n: number }).n;
  }

  getLastRoastAt(): number {
    const row = this.metaGetStmt.get(META_LAST_ROAST_AT) as { value: string } | undefined;
    if (!row?.value) return 0;
    const n = Number(row.value);
    return Number.isFinite(n) ? n : 0;
  }

  setLastRoastAt(ts: number): void {
    this.metaSetStmt.run(META_LAST_ROAST_AT, String(ts));
  }

  removeByIds(ids: number[]): void {
    for (const id of ids) this.removeIdsStmt.run(id);
  }

  isOptedOut(userUid: string): boolean {
    return !!this.isOptedStmt.get(userUid);
  }

  /** Opt a user out of the roast and purge everything they've said. Returns rows removed. */
  optOut(userUid: string): number {
    this.optInStmt.run(userUid);
    return this.purgeStmt.run(userUid).changes;
  }
}

function clampScore(n: number): number {
  if (!Number.isFinite(n)) return 0;
  return Math.max(0, Math.min(10, Math.round(n)));
}

function rowToQuote(r: any): RoastQuote {
  return {
    id: r.id,
    userUid: r.user_uid,
    userName: r.user_name,
    text: r.text,
    createdAt: r.created_at,
    score: r.score ?? null,
    reason: r.reason ?? null,
  };
}

File: crates/mp-db/src/roast.rs
// Copyright (c) 2026 Lane Ambrose
// SPDX-License-Identifier: MIT

use crate::{Database, Result};

#[derive(Debug, Clone)]
pub struct RoastQuote {
    pub id: i64,
    pub user_uid: String,
    pub user_name: String,
    pub text: String,
    pub created_at: i64,
    pub score: Option<i64>,
    pub reason: Option<String>,
}

pub struct RoastStore<'a> {
    pub(crate) db: &'a Database,
}

impl RoastStore<'_> {
    pub fn add(&self, user_uid: &str, user_name: &str, text: &str) -> Result<()> {
        let now = now_ms();
        self.db.with_conn(|conn| {
            conn.execute(
                "INSERT INTO roast_quotes (user_uid, user_name, text, created_at) VALUES (?1, ?2, ?3, ?4)",
                rusqlite::params![user_uid, user_name, text, now],
            )?;
            Ok(())
        })
    }

    pub fn has_recent_duplicate(&self, user_uid: &str, text: &str, within_ms: i64) -> Result<bool> {
        let since = now_ms() - within_ms;
        self.db.with_conn(|conn| {
            let n: i64 = conn.query_row(
                "SELECT COUNT(*) FROM roast_quotes WHERE user_uid = ?1 AND text = ?2 AND created_at > ?3",
                rusqlite::params![user_uid, text, since],
                |r| r.get(0),
            )?;
            Ok(n > 0)
        })
    }

    pub fn ungraded(&self, limit: i64) -> Result<Vec<RoastQuote>> {
        self.db.with_conn(|conn| {
            let mut stmt = conn.prepare(
                "SELECT id, user_uid, user_name, text, created_at, score, reason
                 FROM roast_quotes WHERE score IS NULL ORDER BY created_at ASC LIMIT ?1",
            )?;
            let rows = stmt.query_map(rusqlite::params![limit], row_to_quote)?;
            Ok(rows.filter_map(|r| r.ok()).collect())
        })
    }

    pub fn set_grade(&self, id: i64, score: i64, reason: &str) -> Result<()> {
        let score = score.clamp(0, 10);
        let reason = if reason.len() > 280 { &reason[..280] } else { reason };
        self.db.with_conn(|conn| {
            conn.execute(
                "UPDATE roast_quotes SET score = ?1, reason = ?2 WHERE id = ?3",
                rusqlite::params![score, reason, id],
            )?;
            Ok(())
        })
    }

    pub fn top(&self, limit: i64) -> Result<Vec<RoastQuote>> {
        self.db.with_conn(|conn| {
            let mut stmt = conn.prepare(
                "SELECT id, user_uid, user_name, text, created_at, score, reason
                 FROM roast_quotes WHERE score IS NOT NULL
                 ORDER BY score DESC, created_at DESC LIMIT ?1",
            )?;
            let rows = stmt.query_map(rusqlite::params![limit], row_to_quote)?;
            Ok(rows.filter_map(|r| r.ok()).collect())
        })
    }

    pub fn stats(&self, min_score: i64) -> Result<(u32, u32, u32)> {
        self.db.with_conn(|conn| {
            let ungraded: i64 = conn.query_row(
                "SELECT COUNT(*) FROM roast_quotes WHERE score IS NULL",
                [],
                |r| r.get(0),
            )?;
            let graded: i64 = conn.query_row(
                "SELECT COUNT(*) FROM roast_quotes WHERE score IS NOT NULL",
                [],
                |r| r.get(0),
            )?;
            let high: i64 = conn.query_row(
                "SELECT COUNT(*) FROM roast_quotes WHERE score IS NOT NULL AND score >= ?1",
                rusqlite::params![min_score],
                |r| r.get(0),
            )?;
            Ok((ungraded as u32, graded as u32, high as u32))
        })
    }

    pub fn is_opted_out(&self, user_uid: &str) -> Result<bool> {
        self.db.with_conn(|conn| {
            let n: i64 = conn.query_row(
                "SELECT COUNT(*) FROM roast_optout WHERE user_uid = ?1",
                rusqlite::params![user_uid],
                |r| r.get(0),
            )?;
            Ok(n > 0)
        })
    }

    pub fn opt_out(&self, user_uid: &str) -> Result<u32> {
        self.db.with_conn(|conn| {
            conn.execute(
                "INSERT OR IGNORE INTO roast_optout (user_uid) VALUES (?1)",
                rusqlite::params![user_uid],
            )?;
            let n = conn.execute(
                "DELETE FROM roast_quotes WHERE user_uid = ?1",
                rusqlite::params![user_uid],
            )?;
            Ok(n as u32)
        })
    }

    pub fn opt_in(&self, user_uid: &str) -> Result<bool> {
        self.db.with_conn(|conn| {
            let n = conn.execute(
                "DELETE FROM roast_optout WHERE user_uid = ?1",
                rusqlite::params![user_uid],
            )?;
            Ok(n > 0)
        })
    }

    pub fn graded_count(&self, min_score: i64) -> Result<u32> {
        self.db.with_conn(|conn| {
            let n: i64 = conn.query_row(
                "SELECT COUNT(*) FROM roast_quotes WHERE score IS NOT NULL AND score >= ?1",
                rusqlite::params![min_score],
                |r| r.get(0),
            )?;
            Ok(n as u32)
        })
    }

    pub fn remove_by_ids(&self, ids: &[i64]) -> Result<u32> {
        if ids.is_empty() {
            return Ok(0);
        }
        self.db.with_conn(|conn| {
            let mut n = 0u32;
            for id in ids {
                n += conn.execute("DELETE FROM roast_quotes WHERE id = ?1", rusqlite::params![id])?
                    as u32;
            }
            Ok(n)
        })
    }

    pub fn last_roast_at(&self) -> Result<i64> {
        self.db.with_conn(|conn| {
            let v: rusqlite::Result<String> = conn.query_row(
                "SELECT value FROM roast_meta WHERE key = 'last_roast_at'",
                [],
                |r| r.get(0),
            );
            Ok(v.ok().and_then(|s| s.parse().ok()).unwrap_or(0))
        })
    }

    pub fn set_last_roast_at(&self, ts: i64) -> Result<()> {
        self.db.with_conn(|conn| {
            conn.execute(
                "INSERT INTO roast_meta (key, value) VALUES ('last_roast_at', ?1)
                 ON CONFLICT(key) DO UPDATE SET value = excluded.value",
                rusqlite::params![ts.to_string()],
            )?;
            Ok(())
        })
    }
}

fn row_to_quote(r: &rusqlite::Row<'_>) -> rusqlite::Result<RoastQuote> {
    Ok(RoastQuote {
        id: r.get(0)?,
        user_uid: r.get(1)?,
        user_name: r.get(2)?,
        text: r.get(3)?,
        created_at: r.get(4)?,
        score: r.get(5)?,
        reason: r.get(6)?,
    })
}

fn now_ms() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use crate::Database;

    #[test]
    fn add_grade_opt_out() {
        let db = Database::open_in_memory().unwrap();
        let s = db.roast();
        s.add("u1", "Bond", "absolutely cringe take").unwrap();
        assert_eq!(s.ungraded(10).unwrap().len(), 1);
        s.set_grade(1, 9, "yikes").unwrap();
        assert_eq!(s.top(5).unwrap()[0].score, Some(9));
        let n = s.opt_out("u1").unwrap();
        assert_eq!(n, 1);
        assert!(s.is_opted_out("u1").unwrap());
    }
}
